ratis-client/src/main/java/org/apache/ratis/client/impl/DataStreamClientImpl.java

```diff
@@ -230,6 +231,14 @@ public DataStreamOutputRpc stream(ByteBuffer headerMessage) {
 
   @Override
   public DataStreamOutputRpc stream(ByteBuffer headerMessage, RoutingTable routingTable) {
+    if (routingTable != null && routingTable.getPrimary() != null) {
+      // Validate that the primary peer is equal to the primary peer passed
+      // by the RoutingTable
+      Preconditions.assertTrue(
+          dataStreamServer.getId().equals(routingTable.getPrimary()),
+          "The primary peer in the routing table is not the same " +
+              "with the one specified in the client");
```

Comment:
Let's include the peer Ids in the message.

Reply:
Thank you for the review. Updated based on the patch diff.